cmd/terminal.go

```diff
@@ -0,0 +1,367 @@
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"strconv"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/vmihailenco/msgpack"
+	"golang.org/x/crypto/ssh/terminal"
+	"golang.org/x/sys/unix"
+	"golang.org/x/term"
+
+	"github.com/mendersoftware/go-lib-micro/ws"
+	wsshell "github.com/mendersoftware/go-lib-micro/ws/shell"
+
+	"github.com/mendersoftware/mender-cli/log"
+)
+
+const (
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+
+	// Time allowed to read the next pong message from the peer.
+	pongWait = 1 * time.Minute
+
+	// protocols
+	httpsProtocol = "https"
+	httpProtocol  = "http"
+	wssProtocol   = "wss"
+	wsProtocol    = "ws"
+)
+
+var terminalCmd = &cobra.Command{
+	Use:   "terminal DEVICE_ID",
+	Short: "Access a device's remote terminal",
+	Args:  cobra.ExactArgs(1),
+	Run: func(c *cobra.Command, args []string) {
+		cmd, err := NewTerminalCmd(c, args)
+		CheckErr(err)
+		CheckErr(cmd.Run())
+	},
+}
+
+func getWebSocketScheme(scheme string) string {
+	if scheme == httpsProtocol {
+		scheme = wssProtocol
+	} else if scheme == httpProtocol {
+		scheme = wsProtocol
+	}
+	return scheme
+}
+
+// TerminalCmd handles the terminal command
+type TerminalCmd struct {
+	server     string
+	skipVerify bool
+	deviceID   string
+	sessionID  string
+	running    bool
+	stop       chan bool
+	err        error
+}
+
+// NewTerminalCmd returns a new TerminalCmd
+func NewTerminalCmd(cmd *cobra.Command, args []string) (*TerminalCmd, error) {
+	server, err := cmd.Flags().GetString(argRootServer)
+	if err != nil {
+		return nil, err
+	}
+
+	skipVerify, err := cmd.Flags().GetBool(argRootSkipVerify)
+	if err != nil {
+		return nil, err
+	}
+
+	return &TerminalCmd{
+		server:     server,
+		skipVerify: skipVerify,
+		deviceID:   args[0],
+		stop:       make(chan bool),
+	}, nil
+}
+
+// Run executes the command
+func (c *TerminalCmd) Run() error {
+	ctx, cancelContext := context.WithCancel(context.Background())
+	defer cancelContext()
+
+	log.Info(fmt.Sprintf("Connecting to the remote terminal of the device %s...", c.deviceID))
+
+	tokenPath, err := getDefaultAuthTokenPath()
+	if err != nil {
+		return errors.Wrap(err, "Unable to determine the auth token path")
+	}
+
+	token, err := ioutil.ReadFile(tokenPath)
+	if err != nil {
+		return errors.Wrap(err, "Please Login first")
+	}
+
+	// get the terminal width and height
+	termID := int(os.Stdout.Fd())
+	termWidth, termHeight, err := terminal.GetSize(termID)
+	if err != nil {
+		return errors.Wrap(err, "Unable to get the terminal size")
+	}
+
+	// connect to the websocket
+	deviceConnectPath := "/api/management/v1/deviceconnect/devices/" + c.deviceID + "/connect"
+	parsedURL, err := url.Parse(c.server)
+	if err != nil {
+		return errors.Wrap(err, "Unable to parse the server URL")
+	}
+
+	scheme := getWebSocketScheme(parsedURL.Scheme)
+	u := url.URL{Scheme: scheme, Host: parsedURL.Host, Path: deviceConnectPath}
```

Comment:
A slight simplification:
```suggestion
	u, err := url.Parse(strings.TrimSuffix(c.server, "/") + deviceConnectPath)
	if err != nil {
		return errors.Wrap(err, "Unable to parse the server URL")
	}
	u.Scheme = strings.Replace(u.Scheme, "http", "ws", 1)
```
(renders `getWebsocketScheme` unused)